### DISCUSSION TITLE
ExternalImageAsTexture: MultiSampling and SVG support

### DISCUSSION THREAD

**Discussion by shriharirathod** [gpuweb/gpuweb#4403] [Q&A]
1. I'm loading an external image as texture in WebGPU. As per the below validation is mentions multisampling while copyExternalImageToTexture does not support destination texture to have sample count other than 1.
Source: [Validation](https://developer.mozilla.org/en-US/docs/Web/API/GPUQueue/copyExternalImageToTexture#validation)
Is there any way to enable multisampling for textures created using external sources?

2. Rendering textures created out of SVG images are pixelating or losing crispness when sampled using WebGPU.
Same image is rendered perfectly when rendered as "HTMLElement" on canvas.
Tried with a couple of different decoding techniques to extract imageBitmaps out of svg file. Results are similar and SVG image quality is reduced. 
Are there any optimal settings for sampling, copyImageAsTexture and createTexture() to render SVG files perfectly with WEBGpu?

**Comment by Kangz** [comment]
1) At the moment there isn't a way to `copyExternalImageToTexture` to a multisampled texture. Though I'm not 100% sure why we can't do this given that we require the `RENDER_ATTACHMENT` usage. Maybe we could enable this. However...

2) I don't think that doing copyEI2T to a multisampled texture from an SVG would preserve all the different sample values. That's because the browser already did the resolve step of the multisampled rendering of the SVG by the time the copy call is done. What you are seeing here could be the browser re-rendering the SVG but single-sampled. Do you have an example showing a side-by-side comparison of the SVG and the WebGPU texture being with different crispness? It would help as a repro case to know exactly what's going on (likely a Chromium bug but let's double check)

**Reply by greggman** [reply]
I'd also be curious if the results are different drawing SVG to a 2D canvas

**Reply by kainino0x** [reply]
The bitmap used to upload _should_ definitely be rasterized from the SVG with antialiasing enabled. Ideally it will have been rasterized at the resolution of the copy, but this is less likely to be true in practice, so it aliasing can still appear if it was rasterized at a higher resolution and then downsampled using bilinear filtering.

**Reply by kainino0x** [reply]
Additionally I don't think it's feasible to upload an SVG into a multisampled texture. Only some rasterization algorithms use MSAA (with unsmoothed values per sample), while other algorithms generate smoothed results without using MSAA (which gives much higher quality than MSAA). It's possible to render at 4x resolution (supersampled) and assign the results into the 4 samples, but this would be somewhat wrong since the sample centers would be wrong.

Uploading an SVG into a multisampled texture is also of limited use. If you want to read it from a shader, note it can't be used with textureSample, only textureLoad with explicit samples that you have to combine yourself. You could write it into a texture, then use the texture as a render target to render stuff in front of it, but this is also of limited use - more likely you would want to blit it into a render target in the middle of the render pass, instead, with a similar result to SVG being converted into geometry and drawn into the scene.
In this case I would recommend uploading the SVG into a texture that's 4x the size of the render target, and using a fragment shader to blit it in the render pass.

**Reply by shriharirathod** [reply]
@Kangz @greggman I'll try to make a sample example for 1 on 1 comparison and share the results.

@kainino0x Will play around with the texture size as per your recommendation and get back with the results.

Thanks everyone for your answers and suggestions, will try out the suggestions and revert back on thread.
